Ecoglayout

roemervandermeij opened the pull request:
@StolkArjen @robertoostenveld 

There we go! Please find some changes here, that will allow us to generate layouts that can be used in multiplotXXX/etc, with an outline generated using the individual's anatomy. It's shame I cannot often spend time like this on Fieldtrip projects, I always really enjoy it. 

I've added two options to ft_prepare_layout (naming-wise, open to suggestions):
1) cfg.ieegview
2) cfg.ieeganatomy 
 
The first takes a string, being 1 of 6 possible viewpoints, or 'auto' (a crude but effective selection of 1 of the 6). The second takes a pial surface (individual or template; or MRI, but less supported), and generates an outline that is used to give anatomical 'perspective'. 

The changes are safe, as they are nearly all contained with the elseif based on cfg.ieegview.

Please see the below code examples and figs as a showcase. 
Robert, how do you feel about the convention I apply for each viewpoint w.r.t. left/right/top/bottom? 

Together with another pull request I initiated today, the last figure (with the depths) is interactive for each depth!

With respect to the last figure some tweaks are still needed, obviously, for the labels. Right now, I'm not so certain which way to go. 

And Robert, fantastic btw to finally see you in full Professor's garb during Nietsche's defense! Well deserved!


<br/>
<br/>
 
Arjen, if you'd like to try them out prior to the branch being pulled in master, you can do the following:
```
git checkout -b roemertest
git remote add fieldtriproemer https://github.com/roemervandermeij/fieldtrip.git
git pull fieldtriproemer ecoglayout:roemertest
```
  
  
<br/>
<br/>
  
  
##### Examples below use the pipeline's UCI29 data:
```Matlab
datapath = '/Users/roemervandermeij/Work/Data/tmpdata/SubjectUCI29/';
load([datapath 'SubjectUCI29_elec_tal_fr.mat']);
load([datapath 'SubjectUCI29_freq.mat']);
% get IR29 pial surface
mesh = ft_read_headshape({[datapath 'SubjectUCI29_rh.pial'],[datapath 'SubjectUCI29_lh.pial']});
mesh.coordsys        = 'tal'; % should be added by ft_read_headshape?
elec_tal_fr.coordsys = 'tal'; % should already be here?
```

<br/>

#### general showcase, show each individual layout with all electrodes
```matlab
views = {'LSAG','RSAG','SUP','INF','ANT','POST'}; 
for iview = 1:numel(views)
  cfg = [];
  cfg.elec        = elec_tal_fr;
  cfg.ieegview    = views{iview};
  cfg.ieeganatomy = mesh;
  lay = ft_prepare_layout(cfg);
  figure('numbertitle','off','name',views{iview});
  ft_plot_lay(lay,'labelsize',4);
end
```

<img width="1262" alt="screen shot 2017-02-16 at 11 32 08 am" src="https://cloud.githubusercontent.com/assets/3984911/23025249/90f69b16-f45d-11e6-86dd-c66351c58783.png">


<br/>
<br/>

#### generate grid layout, and use it for multiplot
```matlab
cfg = [];
cfg.elec        = elec_tal_fr;
cfg.channel     = {'LPG*','LTG*'};
cfg.ieegview    = 'LSAG';
cfg.ieeganatomy = mesh;
lay = ft_prepare_layout(cfg);
figure('numbertitle','off','name','grid freq');
cfg = [];
cfg.layout       = lay;
cfg.showoutline  = 'yes';
cfg.baseline     = [-3 0];
cfg.baselinetype = 'relchange';
cfg.interactive  = 'yes';
ft_multiplotTFR(cfg,freq)
```

<img width="1132" alt="screen shot 2017-02-15 at 2 27 31 pm" src="https://cloud.githubusercontent.com/assets/3984911/23025259/98fc837a-f45d-11e6-9e7b-966689684218.png">

<br/>
<br/>

#### generate anatomical layout, and then multiplot freq with the anatomical as a central layout with depths belongsides it in individual layouts
```matlab
% first generate anatomical layout
depthnames = {'RAM*','RHH*','RTH*','ROC*','LAM*','LHH*','LTH*'};
cfg = [];
cfg.elec        = elec_tal_fr;
cfg.channel     = depthnames;
cfg.ieegview    = 'SUP';
cfg.ieeganatomy = mesh;
cfg.skipscale   = 'yes';
cfg.skipcomnt   = 'yes';
anatlay = ft_prepare_layout(cfg);
% plot all, with the anatomical in the center
figure('numbertitle','off','name','depth freq');
subplot(3,3,5)
ft_plot_lay(anatlay,'box',false,'mask',false,'labelselect',5);
for idepth = 1:numel(depthnames)
  ind = idepth+(idepth>=5);
  subplot(3,3,ind)
  cfg = [];
  cfg.layout       = '5column'; % horizontal
  cfg.channel      = depthnames{idepth};
  cfg.baseline     = [-3 0];
  cfg.baselinetype = 'relchange';
  cfg.showoutline  = 'yes';
  cfg.showlabels   = 'no';
  cfg.showscale    = 'no';
  cfg.showcomment  = 'no';
  ft_multiplotTFR(cfg,freq);
  title(depthnames{idepth}(1:end-1))
end
```

<img width="1023" alt="screen shot 2017-02-16 at 12 29 14 pm" src="https://cloud.githubusercontent.com/assets/3984911/23025262/a0bc122e-f45d-11e6-98f6-1f27faf409cf.png">

StolkArjen commented:
This is impressive, Roemer!! I couldn't entirely replicate it as 

 ! [rejected]        ecoglayout -> roemertest  (non-fast-forward)
 * [new branch]      ecoglayout -> fieldtriproemer/ecoglayout

.. getting regular topoplots instead of mesh outlines. But this may be due to the above git issue (somebody pull this in!).

Regarding the depth layouts, would it also be possible to plot the 7 electrodes in a row per depth instead of two rows of 4 and 3?

robertoostenveld commented:
The code change is too large to get a quick overview. Let me make some suggestions here, but note that I might want more changes once it is merged. 

First: I like the figures!

Roemer> "how do you feel about the convention I apply for each viewpoint w.r.t. left/right/top/bottom?"

Robert> that should be left/right/anterior/posterior/superior/inferior, consistent with the axes labels in ft_determine_coordsys. There is ambiguity in view-from and view-to. It should be consistent with the MATLAB 3D plot view command, which I believe to indicate "view-from".

I don't like the name of the cfg.ieegview and cfg.ieeganatomy options. It is not ieeg specific, or should not be.

Imagine cfg.channel={'MLO*', 'MRO*'} for the CTF151 system, you should then be able to make a posterior view. Or imagine EEG, you should be able to make a 'inferior' view (where the current default is of course superior, i.e. from the top). 

Regarding cfg.ieeganatomy, why not call it cfg.headshape? That is an option that we use for very similar stuff elsewhere. Again, it should be possible to construct a CTF layout with the outline of a head (or brain) behind it.

We now have cfg.projection in ft_prepare_layout. I think that should be the starting point: the projections you are doing are all orthographic (which is not the default). Why not something like this:

cfg.channel = {'MLO*', 'MRO*'}
cfg.projection='orthographic'
cfg.headshape=some nice surface
cfg.view='posterior'
cfg.grad = an 151 channel grad
layout = ft_prepare_layout(cfg);

and of course similar for a proper iEEG elec structure ;-)

Do you have some example data in a (single) mat file? I would like to make a full-fledged test script that does not only iEEG, but also regular EEG and MEG.

roemervandermeij commented:
@StolkArjen 
Oh yeah, that wouldn't work, sorry. You can do the following, which creates a new branch that is identical to the my branch here:
`git checkout -b newroemertest fieldtriproemer/ecoglayout`
(a new branch is created called newroemertest)
Note that without merging this branch with my additions in a different branch (now in master) the subplots are not interactive. I'll see if I can manage that.

Regarding the 1 row vs 2. Sure! Every layout is possible actually, it's just placed in a subplot. Though 1 row is possible (cfg.layout = '7column'), the downside is that, due to the way the width of subplots is handled, it will make the individual channel plots either very narrow strip (cfg.layout = 'horizontal'), or very small rectangles (cfg.layout = '7column'). The 4/5column situation would be sort of a compromise there. (BTW, there's supposedly an error with cfg.layout = '1column', but I'll look into it). And actually, I'll also try and implement the cfg.layout = Xrow version.

@robertoostenveld 

> The code change is too large to get a quick overview. Let me make some suggestions here, but note that I might want more changes once it is merged.

Sure! That was my thought as well actually. The easiness with which I can add commits to this pull request is well suited for this (which is otherwise a nuisance ;)). Once we are settled on the below, I'll reorganize the code where necessary.

> Robert: that should be left/right/anterior/posterior/superior/inferior, consistent with the axes labels in ft_determine_coordsys. There is ambiguity in view-from and view-to. It should be consistent with the MATLAB 3D plot view command, which I believe to indicate "view-from".

Sure, I'm fine with this. It is now 'view-from'. I'll look into the view command with string input, although I'm against flipping it to make it 'view-to' if that's the case. 

> I don't like the name of the cfg.ieegview and cfg.ieeganatomy options. It is not ieeg specific, or should not be. Imagine cfg.channel={'MLO*', 'MRO*'} for the CTF151 system, you should then be able to make a posterior view. Or imagine EEG, you should be able to make a 'inferior' view (where the current default is of course superior, i.e. from the top).

We could change it to cfg.viewpoint, or even cfg.viewfrom, to also resolve that ambiguity. Or how about cfg.orthoview(point)?

> Regarding cfg.ieeganatomy, why not call it cfg.headshape? That is an option that we use for very similar stuff elsewhere. Again, it should be possible to construct a CTF layout with the outline of a head (or brain) behind it.

Well, right now, it also accepts an MRI, though the outline generated by that is a bit 'noisy'. But those can be split of course. How would an MRI usually be called? I can also just tear MRI support out, as it currently is only supported in 2014b (due to boundary.m), although it should not be too complicated to extend support either. BTW, there is also a cfg.mesh option here, which is used in conjunction with a cfg.image (I think). 

> We now have cfg.projection in ft_prepare_layout. I think that should be the starting point: the projections you are doing are all orthographic (which is not the default). Why not something like this:
cfg.channel = {'MLO*', 'MRO*'}
cfg.projection='orthographic'
cfg.headshape=some nice surface
cfg.view='posterior'
cfg.grad = an 151 channel grad
layout = ft_prepare_layout(cfg);

Sounds good to me. BTW, I'm not entirely sure whether xxx.chanpos or xxx.elecpos (senspos?) should be used. Also keeping in mind if planars are used...?

> Do you have some example data in a (single) mat file? I would like to make a full-fledged test script that does not only iEEG, but also regular EEG and MEG.

Sort off. The above (elec, freq, mesh) is taken from ftp://ftp.fieldtriptoolbox.org/pub/fieldtrip/tutorial/SubjectUCI29.zip, but coordsys is currently not yet present everywhere. 

With respect to supporting non-ieeg things like the CTF helmet, there's still the issue of the coordinate system. Can CTF for example be converted on the fly to talairach? If that is possible, we should do this with ft_convert_coordsys. If not, as I can imagine some anatomy is required, I could also do a `switch coordsys` when computing the rotation matrices, but every system will have to be hard-coded, it's not trivial to do this for every system. On the other hand, note that only the convention regarding the x,y,z axes are important for the rotation matrices, as a 'small' shift of 0,0,0 is inconsequential (iff the surface is in the same convention...).

StolkArjen commented:
Regarding the coordsys:

"
mesh = ft_read_headshape({[datapath 'SubjectUCI29_rh.pial'],[datapath 'SubjectUCI29_lh.pial']});
mesh.coordsys        = 'tal'; % should be added by ft_read_headshape?
"

Should we add a if ft_check_fileextension = .pial, coordsys = 'tal' to ft_read_headshape?

robertoostenveld commented:
@StolkArjen, I don't think you can guarantee that all pial files are 'tal'. We have ft_determine_coordsys and the hascoordsys option in ft_checkdata. Please use those and don't replicate functionality.

@roemervandermeij, I like cfg.viewpoint. And agreed with 'view-from'. The code that "knows" about coordinate systems is already present elsewhere in FT. It is fine if you first restrict it to tal only, we can add others later. Right now there is also the 10-year old rotate option, which relates to this as well (as the z-axis is always up).

for the MRI you can use the cfg.mri option (also used in ft_prepare_sourcemodel). The cfg.headshape and cfg.mri options are mutually exclusive.

roemervandermeij commented:
@robertoostenveld 
So, I'm gonna be merging the relevant code into subfunction sens2lay of ft_prepare_layout, to make the new functionality fall under cfg.projection = 'orthographic' with cfg.viewpoint = XXX. 

The projection itself is now computation of a transformation matrix via viewmtx (very convenient), and applying it with ft_warp_apply. 

Something I think you'll have an opinion on:
There's a simple, very low-level, function called elproj. This does some basic projections, of which 'orthographic' (in this case, simply dropping the Z-axis) is not used in the code-base, and others are relatively often (like 'polar'). Elproj currently takes only a set of coordinates and a projection method, and is agnostic to the coordinate system. However, in order to do left/right/superior/etc orthographic projections, the coordinate system needs to be known in order perform the appropriate transformation (followed by dropping the z-axis). Because of this, it doesn't make sense to add functionality here, but in the above sens2lay (which has the grad/elec structure, and with it coordsys. Obviously, this is not a nice 'neat' solution. However, I don't see a good alternative to adding it to sens2lay. Thoughts?

Additionally:
It's not obvious why there is a default Z-axis rotation of 90 degrees for CTF/BTI/etc. In my current plan, this will only be applied when cfg.viewpoint is not present combined with cfg.projection = orthographic. To be sure, why the rotation?

roemervandermeij commented:
Alright! It should be ready to go for pulling. I didn't have many datasets to test for bugs, but the CTF151 and an EEG 1020 cap went as it should have. Substantial changes to my additions have been made, and some reorganization of the old code. 

The viewpoint-specific orthographic projections should work for any type of sens now, if they are in tal or mni. Adding new coordinate systems is easy, and should probably only take 15min or so (see subfunction _getorthoviewpos_). So, if somebody wants to give me a grad and a brain in a known coordinate system (is CTF a coordinate system in our definitions?), I can add them easily. 

The new functionality should be used as follows:
```matlab
cfg.elec/grad/file = <elec/grad structure or file>
cfg.projection = 'orthographic'
cfg.viewpoint = left/right/superior/inferior/anterior/posterior
cfg.headshape = '<a mesh, brain surface or whatever>' (mutually exclusive with cfg.mri)
cfg.mri         = <an mri, with or without brain field> (mutually exclusive with cfg.headshape)
```

As you'll notice in the commits above, there two **defaults changed**. @schoffelen: could you chime in perhaps?

First, automatic addition of circle+nose and accompanying mask is now only done when ft_senstype = eeg/meg/nirs. Does this capture all of the necessary suspects? It used to be the case that it was added for practically all layouts, the reverse (as it is in the above) makes more sense. 

Secondly, now, in case no mask is present, there will always be a mask added, generated by the convex hull/boundary of the sensor 'boxes'. I think this is useful default functionality, no?

StolkArjen commented:
Can't wait to use it!

robertoostenveld commented on ft_layoutplot.m:
```diff
@@ -101,6 +103,10 @@
 cfg.visible = ft_getopt(cfg, 'visible', 'on');
+cfg.box     = ft_getopt(cfg, 'box', 'on');
+cfg.mask    = ft_getopt(cfg, 'mask', 'on');
+cfg.box     = istrue(cfg.box);
```
the cfg option should not be overwritten with a boolean value.

robertoostenveld commented:
hmm, there were three small conflicts. I resolved them online (i.e. in the github website), and now they have been committed. But I have no idea where the commit went... I cannot write in @roemervandermeij repository, and my commit did not end up in the master branch either.

I will make my own branch, pull the changes and review once more.

robertoostenveld commented:
I looked at the code and documentation. The documentation required quite some work (and still does). The way that the options combine (or not) is unclear. To evaluate whether the existing functionality still works - and whether the new functionality works as it should - we really need a test script. I suggest to make a combined tutorial and test script, based on http://www.fieldtriptoolbox.org/tutorial/layout

Right now we don't have a tutorial/layout directory on the ftp. I will add it and add the input files from the existing tutorial (to start with).

roemervandermeij commented:
I'm sorry there were conflicts, when I finished this, there were none. In the mean time, the only change was a renaming of chanposorg to chanposold?  (I'm surprised it actually detects this as a conflict, as the changes are non-overlapping).

I'm surprised to see actually, that you do have write access to my fork apparently? You merged fieldtrip/master into roemervandermeij/ecoglayout. In the current version of roemervandermeij/ecoglayout (this PR), your conflict related changes are incorporated. 

However, I don't see any changes to the documentation. Did you commit them? If you tell me what's unclear, I'd be happy to elaborate. I can see the use of cfg.headshape/mri is a bit unclear, I'll add more clarity there. But to prevent conflicts I can't do this before you commit your changes.

A tutorial however, I won't have the time to do. Once the pipeline paper is out/close, there will be a tutorial for it, which will include use of the functionality I added. Though I spent quite some time making the functionality applicable in general for any mesh/anatomy as long as the coordsys is both specified and implemented, it is not a priority for me right now to make a tutorial to teach how to use the functionality in general. 

A test script I can make. I don't have access to the test data, but I can use ref_datasets for some. Is Arjen's data already of the dccn storage? Regardless, I will still depend on somebody at the dccn to test the test script.

StolkArjen commented:
Hey Roemer, 

The latest elecs and meshes are in /data/SubjectUCI29 of the dropbox folder. I kept changing files, hence the zipfile hosted at the ftp is slightly outdated. I will indeed add any scripting/code you may have on how to use it in the guideline.

Thanks!

robertoostenveld commented:
the conflicts were minor, not to worry about. I have some changes on my laptop and in a branch. Please don't make changes for the moment until I have sorted those out.

Regarding tutorial/test script: I'll make that. I recently made a new plotting tutorial (not yet uploaded) and can do the same for layout. I am mainly planning to loop through some of the options, like views, rotate, etc. So nothing fancy, but only aimed to cover a wider selection of options. Organizing the cfg options in logical (related) groups requires some thinking anyway, that is what I want to present in the layout tutorial. No fancy iEEG stuff, mostly going over the old stuff (to ensure that nothing broke).

roemervandermeij commented:
Re: okay, I'll wait with making any changes. I'll be happy to add more, just let me know. I added a protected version of the above in roemervandermeij/ecoglayoutorg (prior to merging master) in case you need to redo a merge, conflict, or other git mess.

Re: test script, thanks, much appreciated. Below a list of combinations and when they should work, and an example or two. Please do also note the two defaults I changed wrt to adding an outline/mask, which I do think make sense, quoted here from above:

> First, automatic addition of circle+nose and accompanying mask is now only done when ft_senstype = eeg/meg/nirs. Does this capture all of the necessary suspects? It used to be the case that it was added for practically all layouts, the reverse (as it is in the above) makes more sense.
> 
> Secondly, now, in case no mask is present, there will always be a mask added, generated by the convex hull/boundary of the sensor 'boxes'. I think this is useful default functionality, no?
  
<br>



```matlab
cfg.headshape = <some surface>
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.projection = 'orthographic'
cfg.viewpoint = 'anterior'
```
_WORKS if elec/grad and headshape contain the same coordsys
WORKS if coordsys is either tal or mni_
  <br>

```matlab
cfg.mri = <some mri>
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.projection = 'orthographic'
cfg.viewpoint = 'anterior';
```
_WORKS if elec/grad and mri contain the same coordsys AND mri either has brain field or can be segmented on the fly (other segmentations can be achieved via ft_prepare_mesh and cfg.headshape)
WORKS if coordsys is either tal or mni_
  <br>

```matlab
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.projection = 'orthographic'
cfg.viewpoint = 'anterior'
```
_WORKS if elec/grad has coordsys 
WORKS if coordsys is either tal or mni_
  <br>


```matlab
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.projection = 'orthographic' OR other projection
cfg.viewpoint = []
```
_Uses old orthographic projection (taking only x/y) or other projections in elproj.m_
  <br>



```matlab
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.projection = <some projection>
cfg.channel = <some sub-selection of channels>
cfg.boxchannel = <same sub-selection of channels>
```
_The width/height of the sensor 'boxes' is determined using only the channels in cfg.boxchannel. This defaults to 'all', as it should be for e.g. sensor caps/helmets (sub-selecting channels should not change the layout). For ECoG a sub-selection is necessary, as otherwise the boxes will almost always be uselessly tiny._
  <br>


```matlab
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.projection = 'orthographic'
cfg.viewpoint = 'auto'
```
_WORKS if elec/grad has coordsys
WORKS if coordsys is either tal or mni_
  <br>



```matlab
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.headshape = <some surface> OR cfg.mri = <some mri>
cfg.projection = <other than orthographic>
```
WORKS IMPROPERLY: the outline will not be projected in the same way (which defaults to superior orthographic). I should have supported this from the go, more streamlined that way. If you're done committing I'll change this. 

<br>

```matlab
cfg.elec/grad/elecfile/gradfile = <some elec/grad/file> 
cfg.headshape = <some surface> OR cfg.mri = <some mri>
cfg.projection = 'orthographic'
cfg.viewpoint = 'auto';
```
FAILS on viewpoint = auto for the outline, alas. An unintended consequence of segmenting electrode/outline projection code to allow cfg.headshape/mri in broad scenario's. I have fix in mind, where sens2lay should outputs the determined viewpoint. Ain't pretty but it will work. Can also just prevent auto when using cfg.headshape/mri.

robertoostenveld commented:
_Re: circle+nose and accompanying mask when ft_senstype = eeg/meg/nirs._ 

I don't know how robust the EEG detection case is. I'll try some stuff out.

_Re: in case no mask is present, there will always be a mask added, generated by the convex hull/boundary of the sensor 'boxes'. I think this is useful default functionality, no?_

Boxes (i.e. corners of the boxes) or sensor positions (center of the boxes)?

robertoostenveld commented:
thanks for the suggested test cases, I'll include those.

robertoostenveld commented:
here is a simple example that does not work any more the way it used to:

```
elec = ft_read_sens('easycap-M10.txt');

figure
cfg = [];
cfg.elec = elec;
layout = ft_prepare_layout(cfg);

figure
ft_plot_lay(layout)
```

![screen shot 2017-03-23 at 09 53 58](https://cloud.githubusercontent.com/assets/899043/24239555/adec4fce-0fae-11e7-93d6-70d743717e43.png)

The sens is not detected as EEG, the mask is the choppy one (from the corners of the boxes), the outline is missing. I will change it such that this returns the old layout (with circle).

robertoostenveld commented:
pfew. It was quite some work, but I think I have all old behavior consistently supported with the new functionality. For example


```
cfg = [];
cfg.grad = sens;  % CTF151
cfg.channel = 'MEG';
cfg.mri = segmentedmri; % from beamformer tutorial
cfg.projection = 'orthographic';
cfg.channel = 'MR*';
cfg.viewpoint = 'right';
layout = ft_prepare_layout(cfg);
figure; ft_plot_lay(layout); title(cfg.viewpoint)
```

returns

![screen shot 2017-03-23 at 12 12 12](https://cloud.githubusercontent.com/assets/899043/24245203/2a6bdb74-0fc2-11e7-8169-6348d9833e79.png)

robertoostenveld commented:
let's close the discussion here and start a new one for specific items that still need to be resolved. One of them is to extend the test script (which is the start of improving the tutorial) with intracranial test cases. I'll look into the example data that @StolkArjen pointed out.
